### ISSUE TITLE
linuxmusl builds for Alpine, latest is only 8.7.0? 

### ISSUE THREAD

**Issue by polarathene** [lovell/sharp-libvips#15]
I mistakenly [opened this issue at libvips repo](https://github.com/libvips/libvips/issues/1205). 

Gist of it is an important fix for Alpine was included in 8.7.1(avoids issues writing to /tmp). 

Could a new release be included(alongside sharp 0.21.2 which has a bugfix I'm looking forward to :) )

I know that [this suggestion exists for building libvips on Alpine](http://sharp.pixelplumbing.com/en/stable/install/#alpine-linux)
```
apk add vips-dev fftw-dev build-base --update-cache \
    --repository https://alpine.global.ssl.fastly.net/alpine/edge/testing/ \
    --repository https://alpine.global.ssl.fastly.net/alpine/edge/main
```

However it's not compatible anymore on stable versions of Alpine, you must use alpine:edge. I believe this is due to edge switching to musl-libc while stable remains on musl. If you were building the binaries for Alpine this way previously, then perhaps a newer release on Alpine isn't an option?(well more work to do)

**Comment by lovell** [comment]
Hello, these binaries are provided to help most of the people most of the time. If you need a newer feature of libvips and can't use Alpine Edge then you'll need to install from source.

**Comment by polarathene** [comment]
> If you need a newer feature of libvips

@lovell it's a bugfix not a feature, specifically for Alpine linux when writing to /tmp. Youv'e actually had [an issue open in the past](https://github.com/lovell/sharp/issues/1436#issuecomment-435826166) about it and linked to [your own PR to resolve it](https://github.com/libvips/libvips/pull/1155)([related libvips issue for reference](https://github.com/libvips/libvips/issues/1151)). Also affects macOS users?

I'm not aware of any node alpine base images that offer alpine edge? I'm not sure that maintaining my own node base image to use alpine edge is advisable?

**Comment by polarathene** [comment]
I managed to compile libvips looking elsewhere... and then realized this repo itself automates the building for alpine with docker too.. oops. I was wondering where all the other dependencies were coming from when comparing my libvips build vs the binary download I had of 8.7.0.

It seems that the build script `lin.sh` compares versions to some service and refuses to continue if they're not all at the latest version? You get terminal output about the package version having been superseded, but after it's done all it's checks, it just exits if something was superseded/newer... **adding a helpful message about the fact it did fail the check would be good rather than just exiting silently** :) I had thought it wasn't an issue at first but noticed I had no build.

```
xml2 version 2.9.8 has been superseded by 2.9.9
gsf version 1.14.44 has been superseded by 1.14.45
harfbuzz version 2.1.3 has been superseded by 2.3.0
pango version 1.42.4 has been superseded by 1.43.0
svg version 2.45.0 has been superseded by 2.45.2
```

Ok, so these are distro agnostic and being built from source, would think that building with the versions that were set there vs the latest versions would have been ok? I updated the versions to be the latest, I'm in China via VPN presently, so sometimes this script failed a few times stating something was superseded by a blank version number(null response I guess). 

Attempting to install via the current setup this repo offers has failed several times during due to unreliability of the connection it seems, which means starting the long process from the start...I will build a Dockerfile to utilize layers instead to try avoid the issue..

**UPDATE:** Got a Dockerfile working. `pango` at 1.43.0 requires `meson` to build now, so I guess comment it out of your version check when you do your next round of builds?

**Comment by lovell** [comment]
Glad you got it working.

If you didn't see it, Node is packaged for Alpine Edge - https://pkgs.alpinelinux.org/package/edge/main/x86_64/nodejs

The automated version check exists to ensure the latest security fixes of all dependencies are always included.

Happy to accept a PR with an error message for a failed latest version check if you're able.

**Comment by polarathene** [comment]
@lovell I guess I could try with the alpine package, just found it curious that node base images on docker hub(not just under node's repo) avoid it to build from source, I guess for offering different versions perhaps?

Are you interested in a PR for the Dockerfile I got working? It's got all the dependencies to be downloaded/compiled in their own images/stages, presently using the alpine Dockerfile as a base, but should be possible to use a build arg to adjust that to the others you have for all that use `lin.sh`. I think I can make all the dependencies use ARG too so that only newer versions would get built otherwise the cache could still be used if desired provided no changes to the base image(though for edge you probably would want to get fresh updates on that for as you mention latest security fixes... so not sure how useful of a contribution to this repo it would be, I don't think I could personally have completed the build with my connection without the help of Docker's layers though.

One other possible benefit is Docker has support for other image builders afaik, one which I've read can intelligently build stages in parallel where possible.

**Comment by polarathene** [comment]
Is this repo how you built the current 8.7.0 release for Alpine? Could just be due to trying to use the compiled file in Alpine 3.8 like I have with 8.7.0 rather than migrate to edge, but I got the following error when trying to use it:

```
Error: Error loading shared library libcairo-gobject.so.2: No such file or directory (needed by /site/node_modules/sharp/build/Release/../../ve  ndor/lib/librsvg-2.so.1)
```

That's meant to be part of the package? For librsvg, I have it being built by adding the Pango build(/target contents) which should include Cairo afaik? From what I can tell I don't need to include `/deps` to build any of these?

```Dockerfile
FROM base-image AS builder-svg
COPY --from=builder-gdkpixbuf ${TARGET} ${TARGET}
COPY --from=builder-pango ${TARGET} ${TARGET}
COPY --from=builder-croco ${TARGET} ${TARGET}

RUN set -x \
  && mkdir ${DEPS}/svg \
  && curl -Lks https://download.gnome.org/sources/librsvg/$(echo "$VERSION_SVG" | sed "s/\.[0-9]*$//")/librsvg-${VERSION_SVG}.tar.xz | tar xJC ${DEPS}/svg --strip-components=1

# Optimise Rust code for binary size
RUN set -x \
  && cd ${DEPS}/svg \
  && sed -i "s/debug = true/debug = false\ncodegen-units = 1\nincremental = false\npanic = \"abort\"\nopt-level = ${RUST_OPT_LEVEL:-\"s\"}/" Cargo.toml \
  && ./configure --host=${CHOST} --prefix=${TARGET} --enable-shared --disable-static --disable-dependency-tracking \
     --disable-introspection --disable-tools --disable-pixbuf-loader \
  && make install-strip

# Clear executable bit from librsvg shared library for WSL support
# Alpine doesn't support execstack? (/bin/sh: execstack: not found)
# RUN execstack -c ${TARGET}/lib/librsvg-2.so || true
```

**Comment by polarathene** [comment]
I removed the flag `--disable-gobject`, which resolved the error... however now I have:

```
Error: Error relocating /site/node_modules/sharp/build/Release/../../vendor/lib/libexpat.so.1: getrandom: symbol not found
```

This is the one I remember getting when building libvips(vips-dev) from edge repos and trying to run it on Alpine 3.8, something due to switch to musl-libc in edge I think. And I'm unable to compile librsvg on 3.8 due to Cargo being out of date(doesn't seem like there is an easy way to fix that?).

**Comment by lovell** [comment]
See commit ef0d685 for a few updates/fixes that should probably help you.

The build order of the various dependencies is important due to their interdependence. A multi-layer image may help with performance but I guess a tradeoff would be increased complexity. Running the scripts on a VM hosted outside China might provide a similar performance improvement ;)

If you think things could be simplified via docker ARGs then that might be a good place to start.

**Comment by lovell** [comment]
Binaries for libvips v8.7.4 now available.
